File: Javascript/Testing/scripts/sll.js
const Node = require('./node');

class SLL {
    constructor(){
        this.head = null;
        this.count = 0; 

    }

    // Inserts 
    insertAtHead(val) {
        const newNode = new Node(val, this.head);
        this.head = newNode;
        this.count++ 
        return this 
    } 

    append(val){
        const newNode = new Node(val, null);
        let runner = this.head;

        while (runner.next){
            runner = runner.next; 
        }
        runner.next = newNode; 
        this.count++
        return newNode;
    }

    insertAtIndex(idx, val){
        if (idx < 0 || idx >= this.count) return null;

        let runner = this.head;
        let prev = null;
        const newNode = new Node(val, null);
        
        for (let i = 0; i < this.count; i++){
            if (i == idx) {
                prev.next = newNode; 
                newNode.next = runner 
                return this;
            }
            prev = runner; 
            runner = runner.next; 
        }
    }

    // Getters 
    getByIndex(idx) {
        if (idx < 0) return null;
        if (idx >= this.count) return null;
        let runner = this.head;
        
        for (let i = 0; i < this.count; i++){
            if (i == idx) return runner;
            runner = runner.next;
        }
    }


    // Removers 
    removeAtHead() {
        let temp = this.head;
        this.head = this.head.next;
        temp = null;
        this.count-- 
        return temp;
    }

    pop() {
        if (this.count == 1) return this.removeAtHead(); 

        let runner = this.head; 
        while (runner.next.next) {
            runner = runner.next; 
        }
        let temp = runner.next; 
        runner.next = null; 
        this.count-- 
        return temp;
    }

    removeAtIndex(idx) {
        if (idx < 0 || idx >= this.count) return null  

        if (idx == 0) return this.removeAtHead(); 
        if (idx == this.count - 1) return this.pop(); 

        let runner = this.head; 
        for (let i =0; i < idx-1; i++){
            runner = runner.next; 
        }
        let temp = runner.next; 
        runner.next = runner.next.next 
        temp.next = null; 
        this.count--
        return temp; 

    }

    // Display
    printValues() {
        let output = "";
        let runner = this.head;
        while (runner) {
            output = `${output}${runner.value} -> `
            runner = runner.next;
        }
        console.log(`${output}null`)
    }

    
} 

SLL.fromValues = (...values) => {
    const sll = new SLL();
    for (let i = values.length -1; i >= 0; i--) {
        sll.insertAtHead(values[i])
    }
    return sll;
}

module.exports = SLL;

File: Java/DataStructures/SLL/SLL.java
public class SLL {
    public Node head;
    public int count;
    public SLL(){
        this.head = null;
        this.count = 0;
    }

    public void add(int value) {
        Node newNode = new Node(value);
        if(head == null) {
            head = newNode;
            count += 1;
        } else {
            Node runner = head;
            while(runner.next != null) {
                runner = runner.next;
            }
            runner.next = newNode;
            count += 1;
        }
    }    

    public SLL remove(){
        Node runner = head;
        while (runner.next.next != null) {
            System.out.println(runner);
            runner = runner.next;
        }
        System.out.println("Node to be removed: " + runner.next.value);
        runner.next = null;
        count -= 1;
        return this;
    }

    public Node removeAt(int val) {
        Node runner = head;
        Node prev = head;
        int idx = 0;

        if (val == 0){
            Node temp = head; 
            head = head.next;
            temp = null;
            return runner;
        }
        while (runner.next != null || idx != count ) {
            if (idx == val){
                Node returnNode = runner;
                Node temp = runner.next;
                prev.next = temp;
                runner.next = null;
                return returnNode;
            }
            else {
                prev = runner;
                runner = runner.next;
            }
            idx+=1;
        }
        return runner;
    }

    public Node find(int val) {
        Node runner = head; 
            while (runner.next != null ){
                if (runner.value == val){
                    System.out.println("Found it! " + runner + ":" + " " + runner.value);
                    return runner;
                }
                runner = runner.next;
            }
        System.out.println("We couldn't find the node");
        return runner; 
    }

    public void printValues() {
        Node runner = head;
        while (runner != null) {
            System.out.println(runner.value);
            runner = runner.next;
        }
    }


}